### PR TITLE
1825: start of OR; track laying rules

### PR DESCRIPTION
**Pull request by roseundy**
Implement OR operating order
Implement track laying rules (except for minors w/o home hex cities)

Common file edits:
- Add 'OO' label to 52 tile (only used by 1825) (config/tiles)
- Allow 'blocks_hexes' ability directly on Corporations (game/base, step/tracker)
- Fix issue where TrackAndToken didn't combine Track and Token round_state and setup methods correctly

### CODE REVIEW

File: lib/engine/corporation.rb [hunk 1]
```diff
@@ -255,6 +255,14 @@ def share_percent
       @second_share&.percent || presidents_percent / 2
     end
 
+    def border?
```
Comment (tobymao) [inline lib/engine/corporation.rb:258]: why?

File: lib/engine/corporation.rb [hunk 2]
```diff
@@ -255,6 +255,14 @@ def share_percent
...
+      false
+    end
+
+    def sym
```
Comment (tobymao) [inline lib/engine/corporation.rb:262]: what's this?

File: lib/engine/game/base.rb [hunk 3]
```diff
@@ -2007,6 +2011,10 @@ def init_hexes(companies, corporations)
         end.flatten.compact
       end
 
+      def blocker_companies
```
Comment (tobymao) [inline lib/engine/game/base.rb:2014]: why is this necessary?

File: lib/engine/game/base.rb [hunk 4]
```diff
@@ -1935,7 +1939,7 @@ def game_corporations
 
       def init_hexes(companies, corporations)
         blockers = {}
-        companies.each do |company|
+        blocker_companies.each do |company|
```
Comment (tobymao) [inline lib/engine/game/base.rb:1942]: why don't you just change this to be companies + corporations

File: lib/engine/step/tracker.rb [hunk 5]
```diff
@@ -86,9 +86,9 @@ def lay_tile(action, extra_cost: 0, entity: nil, spender: nil)
         old_tile = hex.tile
         graph = @game.graph_for_entity(entity)
 
-        @game.companies.each do |company|
+        @game.track_blockers.each do |company|
```
Comment (tobymao) [inline lib/engine/step/tracker.rb:89]: maybe just do companies + corporations?

File: lib/engine/corporation.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/corporation.rb:258]: I'm adding the :blocks_hexes ability to corporations (as opposed to just companies). They are now allowed to be blockers. Both Engine::Tile, UI::Blocker and UI::LocationName expect blockers to have a border? method.

File: lib/engine/game/base.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/base.rb:2014]: I did not want to break games that didn't allow corporations to be blockers.

File: lib/engine/game/base.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/base.rb:1942]: See above.

File: lib/engine/step/tracker.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/step/tracker.rb:89]: See above.

File: lib/engine/corporation.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/corporation.rb:262]: See above.

File: lib/engine/game/base.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/base.rb:2014]: I did the same thing for partition_companies months ago.

File: lib/engine/corporation.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/corporation.rb:262]: Again, the UI expects blockers to have a sym method.

File: lib/engine/corporation.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/corporation.rb:258]: All right. I think I can avoid this if I can change Engine::Tile to not call the border? method for blockers. Then border? can be removed on both corporations and companies.

Comment (roseundy) [reply lib/engine/corporation.rb:258]: Or not, I need to understand why the UI does this:
@tile.parts.reject(&:border?).one?

File: lib/engine/corporation.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/corporation.rb:262]: Nevermind. Removed.

File: lib/engine/corporation.rb (hunk 1 above)
Comment (roseundy) [reply lib/engine/corporation.rb:258]: Removed blockers from @parts

File: lib/engine/game/base.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/base.rb:2014]: Removing.

File: lib/engine/game/base.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/base.rb:1942]: Removing.

File: lib/engine/step/tracker.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/step/tracker.rb:89]: Yep. Doing what you suggest.

File: assets/app/view/game/part/blocker.rb [hunk 6]
```diff
@@ -54,7 +54,7 @@ class Blocker < Base
         }.freeze
 
         def preferred_render_locations
-          if @tile.parts.reject(&:border?).one?
+          if @tile.parts.reject(&:border?).empty?
```
Comment (tobymao) [inline assets/app/view/game/part/blocker.rb:57]: tile.parts.none?(:border?)

File: assets/app/view/game/part/location_name.rb [hunk 7]
```diff
@@ -80,7 +80,7 @@ def preferred_render_locations
             return [center, l_up40, l_down40, l_bottom, l_top]
           end
 
-          @tile.parts.reject { |p| p.path? || p.border? }.any? ? [l_down40] : [l_center]
+          @tile.parts.reject { |p| p.path? || p.border? }.any? || @tile.blockers.any? ? [l_down40] : [l_center]
```
Comment (tobymao) [inline assets/app/view/game/part/location_name.rb:83]: parts.any?  {... } || !@tile.blockers.empty?

File: assets/app/view/game/part/blocker.rb (hunk 6 above)
Comment (roseundy) [reply assets/app/view/game/part/blocker.rb:57]: I don't think that is the same function.

Existing: Are there any parts besides borders?
Yours: Are there no borders?

Comment (roseundy) [reply assets/app/view/game/part/blocker.rb:57]: Perhaps !@tile.parts.any? { |p| !p.border? }

Comment (roseundy) [reply assets/app/view/game/part/blocker.rb:57]: OK

Comment (roseundy) [reply assets/app/view/game/part/blocker.rb:57]: But it should be @tile.parts.none? { |p| !p.border? }

Comment (roseundy) [reply assets/app/view/game/part/blocker.rb:57]: @tile.all?(&:border?)

File: assets/app/view/game/part/location_name.rb (hunk 7 above)
Comment (roseundy) [reply assets/app/view/game/part/location_name.rb:83]: Yep

**Review by tobymao — APPROVED** [review]
